File: python/00337_house_robber_III.py

class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return (0, 0)

            left = dfs(node.left)
            right = dfs(node.right)

            rob_this = node.val + left[1] + right[1]

            not_rob_this = max(left) + max(right)

            return (rob_this, not_rob_this)

        return max(dfs(root))

# Time Complexity: O(n); n is the number of nodes in the tree
# Space Complexity: O(h); h is the height of the tree (due to recursion stack)

File: java/00337_house_robber_III.java
class Solution {
    public int rob(TreeNode root) {
        int[] result = dfs(root);
        return Math.max(result[0], result[1]);
    }

    private int[] dfs(TreeNode node) {
        if (node == null) {
            return new int[2];
        }

        int[] left = dfs(node.left);
        int[] right = dfs(node.right);

        int rob = node.val + left[1] + right[1];
        int notRob = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

        return new int[]{rob, notRob};
    }
}

// Time Complexity: O(n); n is the number of nodes in the tree
// Space Complexity: O(h); h is the height of the tree (due to recursion stack)